seccomp/chown.go

```diff
@@ -0,0 +1,205 @@
+//
+// Copyright 2019-2020 Nestybox, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+package seccomp
+
+import (
+	"path/filepath"
+	"syscall"
+
+	"github.com/nestybox/sysbox-fs/domain"
+	"github.com/sirupsen/logrus"
+	"golang.org/x/sys/unix"
+)
+
+// This file contains Sysbox's chown syscall trapping & handling code. The only
+// reason we trap chown (as well as fchown and fchownat) is to prevent chown to
+// /sys inside a sys container from failing. The reason chown to /sys inside a
+// sys container would fail without this code is that /sys is owned by the
+// host's true root, so it shows up as "nobody:nogroup" inside the sys container
+// and thus its ownership can't be changed from within the container. Some apps
+// running inside the container (e.g,. RedHat's RPM package manager) want to
+// chown /sys to root:root, causing the apps to get an EPERM and fail.  As a
+// work-around, Sysbox ignores chown to "/sys" inside the sys container (or in
+// any inner containers). All other chown operations are handled normally by the
+// kernel.
+
+func (t *syscallTracer) ignoreChown(cntr domain.ContainerIface, req *sysRequest, absPath string) bool {
+
+	// Note: we only ignore chown targeting "/sys" directly. We purposely avoid
+	// resolving symlinks to "/sys" because such symlinks are unusual and
+	// resolving them would slow down every chown syscall. This means that if a
+	// user chowns "/sys" by way of one or more symlinks, the syscall will not be
+	// ignored and will thus still fail.
+
+	if absPath != "/sys" {
+		return false
+	}
+
+	// Check if /sys is a sysfs mount. In the rare case where it's not, we can't
+	// ignore the chown.
+
+	mh := t.mountHelper
+	mip, err := NewMountInfoParser(mh, cntr, req.Pid, false)
+	if err != nil {
+		logrus.Errorf("Failed to get mount info while processing fchown from pid %d: %s", req.Pid, err)
+		return false
+	}
+
+	mi := mip.GetInfo(absPath)
+	if mi == nil || mi.FsType != "sysfs" {
+		return false
+	}
+
+	return true
+}
+
+func (t *syscallTracer) processChown(
```

Comment:
Looks like this entire function can be moved to tracer.go as its coupled to syscallTracer types such as sysRequest. Ideally, chown implementation shouldn't need to know anything about our particular syscallTracer implementation.

Reply:
Done.